### DISCUSSION TITLE
getStaticProps SerializableError: Error serializing ... returned from `getStaticProps`

### DISCUSSION THREAD

**Discussion by cezarneaga** [vercel/next.js#18897] [Help]
# Bug report

## Describe the bug

https://codesandbox.io/s/great-hooks-rxi1d?file=/pages/index.js

i reproduced the issue here. i produce correct data (you can see the logs from the function) but the staticProps fail to read them.

## To Reproduce

https://codesandbox.io/s/great-hooks-rxi1d?file=/pages/index.js

## Expected behavior

i expect it to give correct data to components. not sure if i am doing something wrong or is it an issue with the graphql server

## System information

- OS:  macOS,

- Version of Next.js: [e.g. 10.0.1]
- Version of Node.js: [e.g. 15.0.1]

**Comment by olivertilk** [comment]
I got this same error when I was trying to pass an object to a new external function for processing and then trying to use the return value as the new object. The error resolved when I got rid of the external function and moved the code back to the original function.

I see a similar pattern here. You are using an external function (fetchContent) to process an object (result) and then trying to use the return value as the new object (const { items } = ...).

I would try making the fetch call inside of getStaticProps instead of through fetchContent.

**Reply by cezarneaga** [reply]
copy pasting the fetchContent function everywhere seems like a very bad solution. can we not have this as an issue and look for another solution?

**Reply by cezarneaga** [reply]
also it doesn't seem to solve the issue

**Reply by cezarneaga** [reply]
sorry, it does solve the issue. but the copy pasting remains a bummer

**Comment by cezarneaga** [comment]
in this example: https://github.com/vercel/next.js/tree/3f4872aa679c66078a305f9fdbcde97a14f1e755/examples/cms-contentful by @OriginalEXE it also doesn't work.

**Reply by OriginalEXE** [reply]
Hey, thanks for bringing this to my attention. I have submitted a PR that should fix the issue: https://github.com/vercel/next.js/pull/19012

Basically, the problem is that `undefined` is not serializable, so we should fall back to `[]` for collections and `null` for single entries.

**Reply by cezarneaga** [reply]
in my experience this results into nothing being dispalyed

**Reply by OriginalEXE** [reply]
You need to of course handle it on the frontend, in a realistic application you would have some kind of an "empty screen" view, where you display a message that no posts could be found/loaded.

**Reply by cezarneaga** [reply]
the error is at build time. nothing to handle in ui then.

**Reply by OriginalEXE** [reply]
Ah, ok. This is with the contentful cms example and my new patch? Any specific errors?

**Reply by cezarneaga** [reply]
yes. the contentful with graphql. i linked it above. `SerializableError: Error serializing returned from getStaticProps in `

**Reply by olivertilk** [reply]
Even though my solution worked for me, I'm actually still not sure why. The issue for me was not so much that I *want* to display an empty or undefined value. The issue is that for some reason it was seeing 'undefined' instead of the value. I'm starting to think it's just an issue with manipulating objects.

**Reply by cezarneaga** [reply]
i got it to work :) but i am also not sure what i did different :)

**Reply by olivertilk** [reply]
I looked at my code again and I just realized I made a variable naming error. This is why I was getting 'undefined'. And 'undefined' can't be serialized so that's the reason for the error.

Maybe the object property you're getting 'undefined' for just isn't there. Cheers.

**Comment by robin3317** [comment]
I got the same error and here how I fix it. If I return the fetched data directly as props like bellow - it occurs error.
`return {props: res.data}`
But if I return the fetched data like this(as a plain object), it doesn't occur that `SerializableError`.
```
export async function getStaticProps(context) {
  const res = await axios.get(
    'api_url'
  )
  console.log(res.data)

  return {
    props: {
      departments: res.data,
    },
  }
}
````

**Reply by oswaldosan** [reply]
nice!, this works perfectly

**Reply by seiwonpark** [reply]
Thank you so much! great explanation!

**Comment by PriyaBihani** [comment]
I was getting the same error, but in my case it was due to the timestamps. Check if your object contains timpestamps for createdAt and updatedAt property. Convert them into a string.

**Reply by alvara** [reply]
@PriyaBihani Thank you, this ended up being the solution for me!

**Comment by alicerocheman** [comment]
I'm getting the same error, with a prop that's just an enum value with no data fetched:
```ts
return {
  props: {
    pageName: PageNameEnum.myPageName
  }
}
``
removing the .next folder and starting over fixed it for me.
